File: 382-linked-list-random-node/linked-list-random-node.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
**/

class Solution {
    private List<Integer> list;
    private Random random;

    public Solution(ListNode head) {
        ListNode temp = head;
        list = new ArrayList<>();
        random = new Random();

        while (temp != null) {
            list.add(temp.val);
            temp = temp.next;
        }

   
    }

    public int getRandom() {
        int index = random.nextInt(list.size());
        return list.get(index);
    }
}


/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(head);
 * int param_1 = obj.getRandom();
 */

File: 382-linked-list-random-node/linked-list-random-node.py
#solution 1  - hack  create a list and choose a ranodm index

#solution 2-  using random sampling- this is a special case of 
#where k = 1 (ie length of reservior)

#this is solution to the followup , that is when list length now known and when additional space not used 



# t n*m (n = num of calls to getrandom, m = len of linked list  )
# s  1





class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.head  = head
    def getRandom(self) -> int:
        res = 0
        count =1 
        temp = self.head
        while(temp!=None):
            #we keep the prev element in the reservior iff its prob = (k/k+i)=(1/count) 
            #(here i is count and k=1 )
            #else we need to replace the element
          
            if( random.random() < 1/count ):
                res = temp.val
          
            temp = temp.next
            count+=1
        return res


